pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -554,6 +570,106 @@ func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.Serv
 	return out, nil
 }
 
+// getProxyServiceInstancesFromMetadata retrieves ServiceInstances using proxy Metadata rather than
+// from the Pod. This allows retrieving Instances immediately, regardless of delays in Kubernetes.
+// If the proxy doesn't have enough metadata, an error is returned
+func (c *Controller) getProxyServiceInstancesFromMetadata(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
+	if len(proxy.WorkloadLabels) == 0 {
+		return nil, fmt.Errorf("no workload labels found")
+	}
+
+	if proxy.Metadata[model.NodeMetadataClusterID] != c.ClusterID {
+		return nil, fmt.Errorf("proxy is in cluster %v, but controller is for cluster %v", proxy.Metadata[model.NodeMetadataClusterID], c.ClusterID)
+	}
+
+	// Create a pod with just the information needed to find the associated Services
+	dummyPod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: proxy.ConfigNamespace,
+			Labels:    proxy.WorkloadLabels[0],
+		},
+	}
+
+	// Find the Service associated with the pod.
+	svcLister := listerv1.NewServiceLister(c.services.informer.GetIndexer())
+	services, err := svcLister.GetPodServices(dummyPod)
+	if err != nil {
+		return nil, fmt.Errorf("error getting instances: %v", err)
+
+	}
+	if len(services) == 0 {
+		return nil, fmt.Errorf("no instances found: %v ", err)
+	}
+
+	// This is a reference to a port by name. We need to do a lookup from the ports map
+	// podPorts must first be unmarshalled, then we can do a lookup.
+	var podPorts []*v1.ContainerPort
+	// We only need the ports if they do a port reference by name, so we don't need to fail yet if
+	// port metadata is not provided
+	if _, f := proxy.Metadata[model.NodeMetadataPodPorts]; f {
+		if err := json.Unmarshal([]byte(proxy.Metadata[model.NodeMetadataPodPorts]), &podPorts); err != nil {
+			return nil, err
+		}
+	}
+
+	out := make([]*model.ServiceInstance, 0)
+	for _, svc := range services {
+		svcAccount := proxy.Metadata[model.NodeMetadataServiceAccount]
+		hostname := kube.ServiceHostname(svc.Name, svc.Namespace, c.domainSuffix)
+		c.RLock()
+		modelService, f := c.servicesMap[hostname]
+		c.RUnlock()
+		if !f {
+			return nil, fmt.Errorf("failed to find model service for %v", hostname)
+		}
+		for _, port := range svc.Spec.Ports {
+			svcPort, f := modelService.Ports.Get(port.Name)
+			if !f {
+				return nil, fmt.Errorf("failed to get svc port for %v", port.Name)
+			}
+			targetPort, err := findPortFromMetadata(port, podPorts)
+			if err != nil {
+				return nil, fmt.Errorf("failed to find target port for %v: %v", proxy.ID, err)
+			}
+			// Construct the ServiceInstance
+			out = append(out, &model.ServiceInstance{
+				Endpoint: model.NetworkEndpoint{
+					Address:     proxy.IPAddresses[0],
+					Port:        targetPort,
+					ServicePort: svcPort,
+					Network:     c.endpointNetwork(proxy.IPAddresses[0]),
+					Locality:    util.LocalityToString(proxy.Locality),
```

Comment:
@howardjohn This is a regression, we can never get the real locality info from the node labels now.